design/httpproxy-prefix-rewrite.md

```diff
@@ -0,0 +1,262 @@
+
+* ***Path component '/' ambiguity.***
+  This proposal does nothing to help users with the problem of
+  specifying trailing '/' characters correctly. If the prefix doesn't
+  end in a '/', then using a replacement of '/' can result in a
+  final path with a leading '//'.
+
+  *TODO(jpeach): What does Envoy do with an empty replacement?*
+
+* ***Multiple replacements.***
+  Rather than allowing a `ReplacementPrefix` slice, we could
+  obtain similar semantics by requiring the user to always
+  specify separate leaf proxies for each rendered prefix. This
+  locally simplifies the API, but arguably forces more complex
+  deployment topologies. If we can expose richer status
+  information, then I think the additional API complexity is
+  preferable.
+
+## Examples
+
+Rewrite a single prefix:
+
+```
+---
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: httpbin
+spec:
+  virtualhost:
+    fqdn: httpbin.jpeach.org
+  routes:
+  - conditions:
+    - prefix: /
+    services:
+    - name: httpbin
+      port: 80
+    pathRewrite:
+      replacePrefix:
+      - replacement: /my/great/replacement
+```
+
+Rewrite multiple prefixes to a single canonical prefix:
+
+```
+---
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: httpbin
+spec:
+  virtualhost:
+    fqdn: httpbin.jpeach.org
+  routes:
+  - conditions:
+    - prefix: /v1/
+    - prefix: /v2/
+    services:
+    - name: httpbin
+      port: 80
+    pathRewrite:
+      replacePrefix:
+      - replacement: /v3/
+```
+
+Rewrite multiple included proxies:
+
+```
+---
+# Create a root proxy that includes a proxy on the /api/ path:
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: artifactory
+spec:
+  virtualhost:
+    fqdn: artifactory.jpeach.org
+  includes:
+  - name: artifactory-api
+    conditions:
+    - prefix: /
+
+---
+# Now split the root into 2 proxies for /v1/ and /v2/.
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: artifactory-api
+spec:
+ includes:
+  - name: artifactory-v1
+    conditions:
+    - prefix: /v1/token/
+  - name: artifactory-v2
+    conditions:
+    - prefix: /v2/token/
+
+---
+# Now, it turns out we don't need a separate proxy for /v1/ any
+# more, so we can send that traffic over to /v2/
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: artifactory-v1
+spec:
+ includes:
+  - name: artifactory-v2
+
+---
+# Now, rewrite the token path for different versions. We configure
+# multiple services to justify specifying this as a single proxy
+# with multiple replacements.
+---
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: artifactory-v2
+spec:
+  routes:
+    services:
+    - name: artifactory-prod
+      port: 80
+      weight: 95
+    - service: artifactory-dev
+      port: 80
+      weight: 4
+    - service: artifactory-sampler
+      port: 80
+      weight: 1
+      mirror: true
+    pathRewrite:
+      replacePrefix:
+      - prefix: /v1/token/
```

Comment:
I think this is too complicated. _iff_ users have this usecase they can solve this using two different routes or httpproxy documents. If they ask for the ability to rewrite based on the resulting path we could add this later, but I don't think its necessary to replace the missing functionality that we left out of HTTPProxy.

Reply:
Yeh, this example did get a bit too complicated :)

However, as it stands leaf `HTTPProxy` documents can be included from multiple roots or can (I think) be included multiple times from the same root with different routing prefixes. Another perspective is to think of the `Prefix` as defending against unexpected multiple inclusion. Since the replacement has a statically known prefix, if the proxy is included again, a unexpected replacement can't happen. I don't think it's a lot of additional complexity to spec the `Prefix` field to accommodate this.

I touched on this under the heading `Multiple replacements`. Using separate `HTTPProxy` documents requires the user to copy changes to multiple documents, which increases the complexity of their deployments.